yellowbrick/text/postag.py

```diff
@@ -114,251 +115,381 @@ def __init__(
-            "INTJ": "interjection",
-            "SYM": "symbol",
-        }
-
-        for tagged_doc in X:
-            for tagged_sent in tagged_doc:
-                for _, tag in tagged_sent:
-                    if tag == "SPACE":
-                        continue
-                    self.pos_tag_counts_[jump.get(tag, "other")] += 1
-
-    def _handle_treebank(self, X):
+        default_dic = {
+                        "noun": 0,
+                        "verb": 0,
+                        "adjective": 0,
+                        "adverb": 0,
+                        "adposition": 0,
+                        "determiner": 0,
+                        "pronoun": 0,
+                        "conjunction": 0,
+                        "infinitive": 0,
+                        "punctuation": 0,
+                        "number": 0,
+                        "interjection": 0,
+                        "symbol": 0,
+                        "other": 0,
+                    }
+        if self.stack:
+            uni_tag_map={}
+            for i in self.labels_:
+                uni_tag_map[i]=default_dic.copy()
+            return uni_tag_map
+        else:
+            return default_dic
+
+    def _handle_universal(self, X,y):
+            """
+            Scan through the corpus to compute counts of each Universal
+            Dependencies part-of-speech.
+    
+            Parameters
+            ----------
+            X : list or generator
+                Should be provided as a list of documents or a generator
+                that yields a list of documents that contain a list of
+                sentences that contain (token, tag) tuples.
+            """
+            jump = {
+                # combine proper and regular nouns
+                "NOUN": "noun", "PROPN": "noun",
+                "ADJ": "adjective",
+                "VERB": "verb",
+                # include particles with adverbs
+                "ADV": "adverb", "PART": "adverb",
+                "ADP": "adposition",
+                "PRON": "pronoun",
+                "CCONJ": "conjunction",
+                "PUNCT": "punctuation",
+                "DET": "determiner",
+                "NUM": "number",
+                "INTJ": "interjection",
+                "SYM": "symbol",
+            }
+            if self.stack: 
+                for tagged_doc,y in zip(X,y):
+                    for tagged_sent in tagged_doc:
+                        for _, tag in tagged_sent:
+                            if tag == "SPACE":
+                                continue
+                            self.pos_tag_counts_[y][jump.get(tag, "other")] += 1
+            else:
+                for tagged_doc in X:
+                    for tagged_sent in tagged_doc:
+                        for _, tag in tagged_sent:
+                            if tag == "SPACE":
+                                continue
+                            self.pos_tag_counts_[jump.get(tag, "other")] += 1
```

Comment:
I really appreciate your efforts at ensuring the control flow logic works correctly! The issue here is that the control flow happens too high in the logic -- leading to a lot of repeated code. Consider the following control flow logic instead:

```python
for idx, tagged_doc in enumerate(X):
    for tagged_sent in tagged_doc:
        for _, tag in tagged_sent:

            if tag == "SPACE":
                continue

            # get counter correct counter from arguments 
            counter = self.pos_tag_counts_[y[idx]] if self.stack else self.pos_tag_counts_ 
            counter[jump.get(tag, "other")] += 1
```

Note that this results in the same number of loops as the original code and your modification and is much more readable at the cost of potentially tracking an unnecessary `idx` in the case of `self.stack is False`. This cost is more than worth it though, as it makes it easier to modify this code in the future and reduces the number of bugs.

Reply:
I will make it a point to avoid building duplicate code. I understand that it makes the code look more complex for the new developers.